
# 1.读取数据+基因ID转换
rm(list = ls())  #q清除环境变量
library(tibble)
options(stringAsFactors = F)  #默认不将字符型变量转换为因子型变量
a = read.table("count.txt",header = T) #读取数据
meta = a[,1:6]##前6列属于表达矩阵的meta信息
exprset = a[,7:ncol(a)]##第7列之后属于真正的表达矩阵
rownames(exprset) <- a$Geneid #将基因名作为行名
colnames(exprset) = gsub(".bam","",colnames(exprset)) #去除后缀.bam

library(org.Hs.eg.db) #下载人类基因组注释
g2s <- unique(toTable(org.Hs.egSYMBOL))#得到gene_id和symbol列  	#org.At.tair.db   拟南芥包
head(g2s)
g2e <- unique(toTable(org.Hs.egENSEMBL)) #得到gene_id 和ensembl_id列
head(g2e)
s2e=merge(g2e,g2s,by='gene_id')#整合g2e和g2s，以gene_id 列为标准
table(a$Geneid %in% s2e$ensembl_id) #查看a中的geneid有多少在s2e中

#b = read.csv('SraRunTable.txt',header = T)

symbol <- g2s[match(a$Geneid,s2e$ensembl_id),"symbol"] 
#a$Geneid  在s2e$ensembl_id 中位置，并且是取symbol列的值

table(duplicated(symbol))# 查看重复基因
###使用aggregate根据symbol列中的相同基因进行合并 
counts <- aggregate(exprset, by=list(symbol), FUN=sum)
rownames(counts)=counts[,1]
counts <- counts[,-1]
#counts <- column_to_rownames(counts,'Group.1')

#fix(counts)
#group_list = c('','','','',)
# gl <- data.frame(row.names=colnames(counts),
#                  group_list=group_list)
#counts_1=counts[,c(1,5,9,2,6,10,3,7,11,4,8,12)]

save(counts_1,file="E:\\360MoveData\\Users\\admin\\Desktop\\生信\\R\\rnaseq\\
     1.count.Rdata")


# 2.相关性分析
library(FactoMineR)
library(factoextra)  
library(tidyverse) # ggplot2 stringer dplyr tidyr readr purrr  tibble forcats
library(RColorBrewer)
#筛选出至少在重复样本数量内的表达量counts大于1的行（基因）
##保留较高表达量的基因
keep_feature <- rowSums(counts_1>1) >= 3
table(keep_feature)
counts_filt <- counts_1[keep_feature, ] #替换counts为筛选后的基因矩阵


##转换成TPM
# 注意需要转化的是未经筛选的counts原始矩阵
### 从featurecounts 原始输出文件counts.txt中提取Geneid、Length(转录本长度)，计算tpm
geneid_efflen <- subset(a,select = c("Geneid","Length"))
colnames(geneid_efflen) <- c("geneid","efflen")  
efflen <- geneid_efflen[match(rownames(exprset1),
                              geneid_efflen$geneid),
                        "efflen"]
### 计算 TPM 公式
 #TPM (Transcripts Per Kilobase Million)  每千个碱基的转录每百万映射读取的Transcripts
counts2TPM <- function(count=exprset1, efflength=efflen){
   RPK <- count/(efflength/1000)   #每千碱基reads (reads per kilobase) 长度标准化
   PMSC_rpk <- sum(RPK)/1e6        #RPK的每百万缩放因子 (“per million” scaling factor ) 深度标准化
   RPK/PMSC_rpk                    
 }  
tpm <- as.data.frame(apply(exprset1,2,counts2TPM))
 colSums(tpm)
#dat <- as.data.frame(log2(edgeR::cpm(counts_filt)+1))# 转换成cpm 也可以，但最好还是TPM
dat <- as.data.frame(tpm+1))

##也可以用DESeq2_normalize   rld
dds <- DESeqDataSetFromMatrix(countData = counts,
                                colData = gl,
                                design = ~ group_list)
rld <- rlog(dds, blind=FALSE)
write.table(assay(rld),  file="Deseq2_rld.txt", sep="\t", quote=F, col.names=NA)
dat <- as.data.frame(assay(rld))     

# 箱型图
boxplot(dat, ylab="dat", main=" normalized data ",
        outline = F, notch = F)

#corrplot
library(corrplot)
corrplot(cor(dat))

# 热图    强烈推荐！！！！！！！
library(pheatmap)
##scale 归一化，让差异性更大一点
pheatmap(scale(cor(log2(counts_filt + 1))),
         angle_col=45,
         fontsize=7,
         annotation_col = gl )

#hclust
##配置参数
nodePar = list(lab.cex = 0.6,pch = c(NA,19),
               cex = 0.7,col = "blue")#节点标签的大小将被调整为默认大小的 60%,
#一个节点（NA）将不绘制点，而第二个节点将使用点类型 19;
#节点的大小将被调整为默认大小的 70%;节点为蓝色
##pch（符号）中设置了两个参数，是因为树形图中有两处用到了符号。NA即不适用符号。
hc = hclust(dist(t(log2(counts_filt+1))))
##先转置再用dist函数计算行之间的距离关系，再使用hclust进行聚类。
par(mar=c(5, 1, 1, 5)+0.1)##设置绘图区距离边界（下，左，上，右）的距离。
#png('hclust.png',res=120)
plot(as.dendrogram(hc),nodePar = nodePar,horiz = TRUE,center = TRUE,
     type = "rectangle")#horiz = TRUE：绘制水平方向的树状图。
#center = TRUE：在水平方向上居中树状图。
#type = "rectangle"：指定节点的形状为矩形。
#dev.off()

#PCA分析  查看实验和对照组情况
##若用TPM
dat.pca <- PCA(t(dat) , graph = F) 
pca <- fviz_pca_ind(dat.pca,
                    title = "Principal Component Analysis",
                    legend.title = "Groups",
                    geom.ind = c("point", "text"), 
                    pointsize = 1.5,
                    labelsize = 4,
                    col.ind = group_list, # 分组上色
                    axes.linetype=NA,  # remove axeslines
                    mean.point=F#去除分组中心点
                    ) + 
  coord_fixed(ratio = 1)+  #坐标轴的纵横比
  xlab(paste0("PC1 (",round(percentVar[1,'variance.percent'],1),"%)")) +
  ylab(paste0("PC2 (",round(percentVar[2,'variance.percent'],1),"%)"))
ggsave(pca, filename = 'check_PCA.pdf',width = 7.5,height =6)

##若用DESeq2_normalize   rld   还可使用DESeq2自带函数
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = counts_filt,colData = gl,
                              design = ~ group_list)
dds = DESeq(dds)
rld <- rlog(dds, blind=FALSE)
pca <- plotPCA(rld, ntop = 500, intgroup=c("group_list"))
ggsave(pca, filename = 'check_PCA.pdf',width = 7.5,height =6)


# 3.差异化分析
## 方法1：DEseq2：必须是整数   around()  不能用归一化后的数据
## 如果是多样本，一起比较，然后用result()得到自己感兴趣的比较
##设定 实验组exp / 对照组ctr
exp="trt"
ctr="untrt"
library(DESeq2)
dds$group_list <- relevel(dds$group_list, ref = ctr) #指定 control group
keep <- rowSums(counts(dds)) >= 1.5*ncol(counts)  #Pre-filtering ，过滤低表达基因
#counts()函数，获取dds中的矩阵  rowsums()函数：计算每行的总和
dds <- dds[keep,] 
dds <- DESeq(dds,quiet = F) 
resultsNames(dds)  #可以看出比较的组
res <- results(dds,contrast=c("group_list", exp, ctr))  #指定提取为exp/ctr结果
#res <- results(dds, name="condition_treated_vs_untreated")
#res <- results(dds, contrast=c("condition","treated","untreated"))
resOrdered <- res[order(res$padj),]  #order根据padj从小到大排序结果 
tempDEG <- as.data.frame(resOrdered)
DEG_DEseq2 <- na.omit(tempDEG)

## 方法2：EBSeq   RSEM Cuffinks 得到count   不需要归一化   
library(EBSeq)
data(GeneMat)
str(GeneMat)
Sizes=MedianNorm(GeneMat) #中值归一化
Conditions=as.factor(rep(c("C1","C2"),each=5))
EBOut=EBTest(Data=GeneMat,
Conditions=as.factor(rep(c("C1","C2"),each=5)),sizeFactors=Sizes, maxround=5)
EBDERes=GetDEResults(EBOut, FDR=0.05)
str(EBDERes$DEfound)
GeneFC=PostFC(EBOut)
str(GeneFC)
data_EBSeq=as.data.frame(GeneFC)



## 火山图
library(ggplot2)
library(pheatmap)
##筛选条件设置
log2FC_cutoff = 1  #2x 或1.5x   
### 样本>=12,可忽略！！11
padj_cutoff = 0.05  #p值0.05
##选取差异分析结果
need_DEG <- DEG_DEseq2[,c(2,6)] #选取log2FoldChange, padj信息
#colnames(need_DEG) <- c('log2FoldChange','padj') 
need_DEG$significance  <- as.factor(ifelse(need_DEG$padj < padj_cutoff &
                    abs(need_DEG$log2FoldChange) > log2FC_cutoff,
ifelse(need_DEG$log2FoldChange > log2FC_cutoff ,'UP','DOWN'),'NOT'))
title <- paste0(' Up :  ',nrow(need_DEG[need_DEG$significance =='UP',]) ,
                '\n Down : ',nrow(need_DEG[need_DEG$significance =='DOWN',]),
                '\n FoldChange >= ',round(2^log2FC_cutoff,3))
g <- ggplot(data=need_DEG, 
            aes(x=log2FoldChange, y=-log10(padj), 
                color=significance)) +
  #点和背景,alpha透明度，size指定圆点的大小
  geom_point(alpha=0.4, size=1) +
  theme_classic()+ #无网格线
  #坐标轴
  xlab("log2 ( FoldChange )") + 
  ylab("-log10 ( P.adjust )") +
  #标题文本
  ggtitle( title ) +
  #分区颜色                  
  scale_colour_manual(values = c('blue','grey','red'))+ 
  #辅助线
  geom_vline(xintercept = c(-log2FC_cutoff,log2FC_cutoff),lty=4,col="grey",lwd=0.8) +
  geom_hline(yintercept = -log10(padj_cutoff),lty=4,col="grey",lwd=0.8) +
  #图例标题间距等设置
  theme(plot.title = element_text(hjust = 0.5), 
        plot.margin=unit(c(2,2,2,2),'lines'), #上右下左
        legend.title = element_blank(), #不显示图例标题
        legend.position="right")  #图例位置
print(g)


## 热图
dat <- log2(edgeR::cpm(counts_filt)+1)
gene_up <- rownames(need_DEG[with(need_DEG,log2FoldChange>log2FC_cutoff & padj<padj_cutoff),])
gene_down <- rownames(need_DEG[with(need_DEG,log2FoldChange< -log2FC_cutoff & padj<padj_cutoff),])
cg <- c(head(gene_up, 50),   #取前50 padj上下调基因名
        head(gene_down, 50))
cg <- na.omit(match(cg,rownames(dat))) 
#match() 函数来查找 cg 中的行名在 dat 数据框的行名中的位置。
#如果找到匹配，就返回行名在 dat 数据框的行号（索引+1），
#如果找不到匹配，就返回 NA
#na.omit()去除NA值
pheatmap(dat[cg,],
         show_colnames =T,
         show_rownames = F,
         #scale = "row",
         fontsize = 7 ,
         cluster_cols = T,
         annotation_col=gl,
         filename = 'heatmap_top50up&down_DEG.pdf')

# KEGG 
library(clusterProfiler)
library(enrichplot)
library(tidyverse)
# library(org.Hs.eg.db)
library(org.Hs.eg.db)
library(DOSE)
library(pathview)
## KEGG
#enrichKEGG()要设置organism = "mmu"，由于其没有readable参数，
#因此之后还要用DOSE包的setReadable进行转化entrez ID为gene Symbol。
gene_up_entrez <- as.character(na.omit(bitr(gene_up, #数据集
                  fromType="SYMBOL", #输入格式
      toType="ENTREZID", # 转为ENTERZID格式
                  OrgDb="org.Hs.eg.db")[,2])) #"org.Hs.eg.db" "org.Mm.eg.db"
gene_down_entrez <- as.character(na.omit(bitr(gene_down, #数据集
                    fromType="SYMBOL", #输入格式
                       toType="ENTREZID", # 转为ENTERZID格式
    OrgDb="org.Hs.eg.db")[,2])) #"org.Hs.eg.db" "org.Mm.eg.db
gene_diff_entrez <- unique(c(gene_up_entrez ,gene_down_entrez ))# 删除重复值

kegg_enrich_results <- enrichKEGG(gene  = gene_up_entrez,
                                  organism  = "hsa", #物种人类 hsa 小鼠mmu
                                  pvalueCutoff = 0.05,
                                  qvalueCutoff = 0.2)
kk_read <- DOSE::setReadable(kegg_enrich_results, 
                             OrgDb="org.Hs.eg.db", 
                             keyType='ENTREZID')#ENTREZID to gene Symbol
write.csv(kk_read@result,'KEGG_gene_up_enrichresults.csv') 
## GO
go_enrich_results <- enrichGO(gene = gene_up_entrez,
                              OrgDb = "org.Hs.eg.db",
                              ont   = "ALL"  ,     #One of "BP", "MF"  "CC"  "ALL" 生物过程（BP）、分子功能（MF）和细胞组分（CC）
                              pvalueCutoff  = 0.05,
                              qvalueCutoff  = 0.2,
                              readable      = TRUE)
save(go_enrich_results, file ='GO_gene_up_enrichresults.Rdata')
browseKEGG(kegg_enrich_results, 'mmu04310') #网页查看通路

#KEGG GO 通路可视化
#KEGG GO 通路可视化
##pathview ——KEGG通路可视化

#构建含log2FC信息的genelist
genelist <- DEG_DEseq2[,2] 
names(genelist) <- rownames(DEG_DEseq2)
genelist <- unique(genelist)
#genelist ID转化
genelist_entrez <- genelist
names(genelist_entrez) <- bitr(names(genelist),
                               fromType="SYMBOL",toType="ENTREZID", 
                               OrgDb="org.Hs.eg.db")[,2]
##查看与选择所需通路
kk_read@result$Description[1:10] #查看前10通路
i=3 
select_pathway <- kk_read@result$ID[i] #选择所需通路的ID号

pathview(gene.data     = genelist_entrez,
         pathway.id    = select_pathway,
         species       = 'hsa' ,      # 人类hsa 小鼠mmu 
         kegg.native   = T,# TRUE输出完整pathway的png文件，F输出基因列表的pdf文件 
         new.signature = F, #pdf是否显示pathway标注
         limit         = list(gene=2.5, cpd=1) #图例color bar范围调整 
)

pathview(gene.data     = genelist_entrez,
         pathway.id    = select_pathway,
         species       = 'hsa' ,      # 人类hsa 小鼠mmu 
         kegg.native   = F,# TRUE输出完整pathway的png文件，F输出基因列表的pdf文件 
         new.signature = F, #pdf是否显示pathway标注
         limit         = list(gene=2.5, cpd=1)) #图例color bar范围调整

## GO富集结果的有向无环图
#(参数只能 ont=bp)
library(GOplot)
plotGOgraph(go_enrich_results)
go_enrich_results <- enrichGO(gene = gene_up_entrez,
                              OrgDb = "org.Hs.eg.db",
                              ont   = "BP"  ,     #One of "BP", "MF"  "CC"  "ALL" 
                              pvalueCutoff  = 0.05,
                              qvalueCutoff  = 0.2,
                              readable      = TRUE)
go_enrich_results=na.omit(go_enrich_results)
gop <- goplot(go_enrich_results, showCategory = 10)
#goplot(go_enrich_results, showCategory = 10)    #跑不出来
# barplot  最常用
barplot(go_enrich_results, font.size=14, showCategory=10)+
  theme(plot.margin=unit(c(1,1,1,1),'lines'))    #ALL
#在不同的 "ONTOLOGY" 类别下，绘制多个分面的条形图
barplot(go_enrich_results, split= "ONTOLOGY", font.size =14)+
  facet_grid(ONTOLOGY~., scale="free")+     
  theme(plot.margin=unit(c(1,1,1,1),'lines'))
#ggsave(barp,filename = 'barp.pdf', width=10, height=10)
# dotplot 最常用
dotplot(go_enrich_results, showCategory = 10)    #ALL
#在不同的 "ONTOLOGY" 类别下，绘制多个分面的散点图
dotplot(go_enrich_results,font.size =8,split = 'ONTOLOGY')+
  facet_grid(ONTOLOGY~., scale="free")+     
  theme(legend.key.size = unit(10, "pt"),#调整图例大小
        plot.margin=unit(c(1,1,1,1),'lines'))#调整四周留白大小

# cnetplot——Gene-Concept Network   
cnetp1 <- cnetplot(go_enrich_results,  foldChange = genelist,
                   showCategory = 6,
                   colorEdge = T,
                   node_label = 'all',
                   color_category ='steelblue')
cnetp2 <- cnetplot(go_enrich_results,  foldChange = genelist,
                   showCategory = 6,
                   node_label = 'gene',
                   circular = T, 
                   colorEdge = T)
ggsave(cnetp1,filename ='cnetplot.pdf', width =12,height =10)
ggsave(cnetp2,filename = 'cnetplot_cir.pdf', width =15,height =10)
# emapplot ——Enrichment Map
#heatplot: Heatmap-like functional classification
#upsetplot——不同富集通路间的重叠基因数量
#treeplot——富集结果术语的层次聚类与高频词标记

# GSEA——基因集富集分析
options(stringsAsFactors = F)
library(org.Hs.eg.db)
library(clusterProfiler)
library(enrichplot)
library(tidyverse)
library(ggstatsplot)
need_DEG <- DEG_DEseq2
need_DEG <- need_DEG[,c(2,5)] #选择log2FoldChange和pvalue（凑成数据框）

colnames(need_DEG) <- c('log2FoldChange','pvalue')
need_DEG$SYMBOL <- rownames(need_DEG)
#转化id  
df <- bitr(rownames(need_DEG), 
           fromType = "SYMBOL",
           toType =  "ENTREZID",
           OrgDb = "org.Hs.eg.db") #人数据库org.Hs.eg.db 小鼠org.Mm.eg.db
need_DEG <- merge(need_DEG, df, by='SYMBOL')  #按照SYMBOL合并注释信息
geneList <- need_DEG$log2FoldChange
names(geneList) <- need_DEG$ENTREZID
geneList <- sort(geneList, decreasing = T)   #从大到小排序
##### gsea富集 ####
KEGG_kk_entrez <- gseKEGG(geneList     = geneList,  #geneList 是个列表
                          organism     = 'hsa', #人hsa 鼠mmu
                          pvalueCutoff = 0.25)  #实际为padj阈值可调整 
KEGG_kk <- DOSE::setReadable(KEGG_kk_entrez, 
                             OrgDb="org.Hs.eg.db",
                             keyType='ENTREZID')#转化id 
GO_kk_entrez <- gseGO(geneList     = geneList,
                      ont          = "ALL",  # "BP"、"MF"和"CC"或"ALL"
                      OrgDb        = "org.Hs.eg.db",#人类org.Hs.eg.db 鼠org.Mm.eg.db
                      keyType      = "ENTREZID",
                      pvalueCutoff = 0.25)   #实际为padj阈值可调整
GO_kk <- DOSE::setReadable(GO_kk_entrez, 
                           OrgDb="org.Hs.eg.db",
                           keyType='ENTREZID')#转化id 

save(KEGG_kk_entrez, GO_kk_entrez, file = "GSEA_result.RData")

## 可视化   
1. barplot
2. cnetplot
3. dotplot
4. emapplot
5. gseaplot
6. goplot
7. upsetplot
kk_gse_cut <- KEGG_kk[KEGG_kk$pvalue<0.05 & KEGG_kk$p.adjust<0.25 & abs(KEGG_kk$NES)>1]
kk_gse_cut_down <- kk_gse_cut[kk_gse_cut$NES < 0,]
kk_gse_cut_up <- kk_gse_cut[kk_gse_cut$NES > 0,]

down_gsea <- kk_gse_cut_down[tail(order(kk_gse_cut_down$NES,decreasing = T),10),]#NES:富集度
up_gsea <- kk_gse_cut_up[head(order(kk_gse_cut_up$NES,decreasing = T),10),]
diff_gsea <- kk_gse_cut[head(order(abs(kk_gse_cut$NES),decreasing = T),10),]

i=2
gseap1 <- gseaplot2(KEGG_kk,
                    up_gsea$ID[i],#富集的ID编号
                    title = up_gsea$Description[i],#标题
                    color = "red", #GSEA线条颜色
                    base_size = 20,#基础字体大小
                    rel_heights = c(1.5, 0.5, 1),#副图的相对高度
                    subplots = 1:3,   #要显示哪些副图 如subplots=c(1,3) #只要第一和第三个图
                    ES_geom = "line", #enrichment score用线还是用点"dot"
                    pvalue_table = T) #显示pvalue等信息
ggsave(gseap1, filename = 'GSEA_up_1.pdf', width =10, height =8)

#绘制下调基因的kegg
i=2
gseap1 <- gseaplot2(KEGG_kk,
                    down_gsea$ID[i],#富集的ID编号
                    title = DOWN_gsea$Description[i],#标题
                    color = "red", #GSEA线条颜色
                    base_size = 20,#基础字体大小
                    rel_heights = c(1.5, 0.5, 1),#副图的相对高度
                    subplots = 1:3,   #要显示哪些副图 如subplots=c(1,3) #只要第一和第三个图
                    ES_geom = "line", #enrichment score用线还是用点"dot"
                    pvalue_table = T) #显示pvalue等信息
ggsave(gseap1, filename = 'GSEA_down_1.pdf', width =10, height =8)
#### 合并 GSEA通路 
gseap2 <- gseaplot2(KEGG_kk,
                   down_gsea$ID,#富集的ID编号
                    title = "DOWN_GSEA_all",#标题
                    color = "red",#GSEA线条颜色
                    base_size = 20,#基础字体大小
                    rel_heights = c(1.5, 0.5, 1),#副图的相对高度
                    subplots = 1:3, #要显示哪些副图 如subplots=c(1,3) #只要第一和第三个图
                    ES_geom = "line",#enrichment score用线还是用点"dot"
                    pvalue_table = T) #显示pvalue等信息
ggsave(gseap2, filename = "GSEA_down_all.pdf",width =12,height =12)


#searank plot 绘制特定基因集的基因排序列表
##绘制up_gsea前3个富集通路
library(cowplot)
library(ggplot2)
pp <- lapply(1:3, function(i) {
  anno <- up_gsea[i, c("NES", "pvalue", "p.adjust")]
  lab <- paste0(names(anno), "=",  round(anno, 3), collapse="\n")
  
  gsearank(KEGG_kk,        #创建图表，
           up_gsea$ID[1], #从id的第一行值开始
           title = up_gsea$Description[i]) + 
    xlab(NULL) +    #去掉x轴标签
    # ylab(NULL) +   #去掉y轴标签
    annotate("text", 10000,
             up_gsea[i, "enrichmentScore"] * .75, 
             label = lab,    #文本注释添加，以lab为注释内容
             hjust=0, vjust=0)
})
rankp <- plot_grid(plotlist=pp, ncol=1)
ggsave(rankp, filename = "gsearank_up.pdf",width=8,height=10)

# ridgeplot山脊图  
library(ggridges)
ridgep <- ridgeplot(KEGG_kk_entrez,
                    showCategory = 15,
                    fill = "p.adjust",
                    core_enrichment = TRUE,
                    label_format = 30, #设置轴标签文字的每行字符数长度，过长则会自动换行。
                    orderBy = "NES",
                    decreasing = F) 
ggsave(ridgep,filename = 'ridgeplot.pdf',width =10,height =10)


# wgcna
options(stringsAsFactors = F)
library(WGCNA)
library(FactoMineR)
library(factoextra)  
library(tidyverse) # ggplot2 stringer dplyr tidyr readr purrr  tibble forcats

#FPKM转换
counts2FPKM <- function(count=exprset1, efflength=efflen){ 
  PMSC_counts <- sum(count)/1e6   #counts的每百万缩放因子 (“per million” scaling factor) 深度标准化
  FPM <- count/PMSC_counts        #每百万reads/Fragments (Reads/Fragments Per Million) 长度标准化
  FPM/(efflength/1000)                                    
}
fpkm <- as.data.frame(apply(exprset1,2,counts2FPKM))
colSums(fpkm)  #每个样本基因fpkm值的总和

# FPKM与TPM的转化
FPKM2TPM <- function(fpkm){
   fpkm/sum(fpkm)*1e6
 }
tpm0 <- as.data.frame(apply(exprset1,2,FPKM2TPM))
colSums(tpm0)
fpkm <- aggregate(fpkm, by=list(symbol), FUN=sum)
library(tibble)
fpkm <- column_to_rownames(fpkm,'Group.1')
fpkm_keep_feature <- rowSums(counts>1) >= 3
table(fpkm_keep_feature)
fpkm_counts_filt <- counts[fpkm_keep_feature, ]

data <- log2(fpkm_counts_filt+1)
fix(data)
### 筛选MAD前5000的基因
keep_data <- data[order(apply(data,1,mad), decreasing = T)[1:5000],]
datTraits <- data.frame(row.names = colnames(data),group=colnames(data))

datExpr0 <- as.data.frame(t(keep_data))
### 判断数据质量--缺失值
gsg <- goodSamplesGenes(datExpr0,verbose = 3) #goodSamplesGenes 该函数用于检测在基因表达数据中好的样本和基因
#erbose = 3 输出详细的信息，包括各种计算过程的提示和结果
gsg$allOK



#go与gesa区别
#GO 确定给定基因集中哪些功能或过程最为显著
#gesa 确定哪些基因集在两个不同条件下的表达模式最为显著  不局限于差异基因，从基因集的富集角度出发

#若不是模式植物，在做go分析时，可用enricher()函数
#enricher(
#  gene,
#  pvalueCutoff = 0.05,
#  pAdjustMethod = "BH",
#  universe,
#  minGSSize = 10,
#  maxGSSize = 500,
#  qvalueCutoff = 0.2,
#  TERM2GENE,
#  TERM2NAME = NA)

##若go和kegg注释文件中 舍弃没有注释的基因
